### PR TITLE
Added user progress functionality

### PR DESCRIPTION
**Pull request by HarisShahidHameed**
Trello link : https://trello.com/c/am5H0bdC/31-progress-of-user-functionality

### CODE REVIEW

File: server/libs/quizlibs.js [hunk 1]
```diff
@@ -76,6 +78,9 @@ const addQuestion = function (query, data) {
 };
 
 const addQuestions = function (query, data) {
+    data.questions.forEach(function(element) {
```
Comment (abdulhaqshah) [inline server/libs/quizlibs.js:55]: use map function

File: server/libs/quizlibs.js [hunk 2]
```diff
@@ -129,7 +134,7 @@ const getQuestionById = function (query,questionId) {
     return new Promise((resolve,reject) => {
         Quiz.find(query).then((quiz) => {
```
Comment (abdulhaqshah) [inline server/libs/quizlibs.js]: findOne

File: server/libs/quizlibs.js [hunk 3]
```diff
@@ -129,7 +134,7 @@ const getQuestionById = function (query,questionId) {
...
             if (quiz) {
-                let question = quiz.find((question) => question._id === questionId);
+                let question = quiz.find((question) => question.questionId === questionId);
```
Comment (abdulhaqshah) [inline server/libs/quizlibs.js]: search with query

File: server/libs/userlibs.js [hunk 4]
```diff
@@ -288,4 +288,178 @@ const uploadPhoto = function(req) {
     })
 };
 
-module.exports = {register, logIn, updateUser, logOut , changePassword, deleteUser, deActivateUser, getUser, uploadPhoto}
\ No newline at end of file
+const updateQuizProgress = function(data) {
+    return new Promise((resolve,reject) => {
+        if (data.answer) {
+            User.updateOne({uuid: data.uuid, "quizProgress.quizId" : data.quizId} , {$push : {"quizProgress.$.attempted" : data.questionId, "quizProgress.$.correct" : data.questionId}}, {new:true})
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: do proper indentation when the line is exceeded from specific limit

File: server/libs/userlibs.js [hunk 5]
```diff
@@ -288,4 +288,178 @@ const uploadPhoto = function(req) {
...
+        }).catch((error) => {
+            reject({
+                status : statusCodes.badRequest,
+                data : error  
+            })
+        })
+    })
+}
+
+const addQuizIdForProgress = function (query, data) {
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: addingQuizToUser

File: server/libs/userlibs.js [hunk 6]
```diff
@@ -288,4 +288,178 @@ const uploadPhoto = function(req) {
...
+        }).catch((error) => {
+            reject({
+                status : statusCodes.badRequest,
+                data : error
+            })
+        })
+    })
+}
+
+const addTutorialIdForProgress = function (query, data) {
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: same as above

File: server/libs/userlibs.js [hunk 7]
```diff
@@ -288,4 +288,178 @@ const uploadPhoto = function(req) {
...
+                status : statusCodes.badRequest,
+                data : error  
+            })
+        })
+    })
+}
+
+const getCompletedQuizzes = function (query, data) {
+    return new Promise((resolve,reject) => {
+        getUser(query).then((user) => {
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: try with query

File: server/libs/userlibs.js [hunk 8]
```diff
@@ -288,4 +288,178 @@ const uploadPhoto = function(req) {
...
+                status : error.status,
+                data : error  
+            })
+        })
+    })
+}
+
+const getCompletedTutorials = function (query, data) {
+    return new Promise((resolve,reject) => {
+        getUser(query).then((user) => {
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: same as above

File: server/routes/quiz.js [hunk 9]
```diff
@@ -29,7 +29,7 @@ router.post('/quiz/addQuestion/:quizId', (req, res) => {
 });
 
 router.post('/quiz/addQuestions/:quizId', (req, res) => {
```
Comment (abdulhaqshah) [inline server/routes/quiz.js]: /quiz/:quizId/question/add

File: server/routes/user.js [hunk 10]
```diff
@@ -75,4 +75,60 @@ router.post('/user/uploadPhoto/:uuid', authenticate, (req,res) => {
     })
 })
 
+router.post('/user/addQuizForProgress', (req,res) => {
```
Comment (abdulhaqshah) [inline server/routes/user.js]: /user/quiz/start

File: server/routes/user.js [hunk 11]
```diff
@@ -75,4 +75,60 @@ router.post('/user/uploadPhoto/:uuid', authenticate, (req,res) => {
...
+    userLibs.addQuizIdForProgress({uuid : req.body.uuid}, req.body).then((user) => {
+        res.status(user.status).send(user);
+    }).catch((error) => {
+        res.status(error.status).send(error);
+    })
+})
+
+router.post('/user/updateQuizProgress', (req,res) => {
```
Comment (abdulhaqshah) [inline server/routes/user.js]: same as above

File: server/routes/user.js [hunk 12]
```diff
@@ -75,4 +75,60 @@ router.post('/user/uploadPhoto/:uuid', authenticate, (req,res) => {
...
+
+router.patch('/user/updateQuizStatus', (req,res) => {
+    userLibs.quizCompleted(req.body).then((user) => {
+        res.status(user.status).send(user);
+    }).catch((error) => {
+        res.status(error.status).send(error);
+    })
+})
+
+router.patch('/user/updateTutorialStatus', (req,res) => {
```
Comment (abdulhaqshah) [inline server/routes/user.js]: id should be in params

**Review by abdulhaqshah — CHANGES_REQUESTED** [review]

File: server/libs/quizlibs.js (hunk 1 above)
Comment (HarisShahidHameed) [reply server/libs/quizlibs.js:55]: Done

File: server/libs/quizlibs.js (hunk 2 above)
Comment (HarisShahidHameed) [reply server/libs/quizlibs.js]: Done

File: server/libs/quizlibs.js (hunk 3 above)
Comment (HarisShahidHameed) [reply server/libs/quizlibs.js]: Done

File: server/libs/userlibs.js (hunk 4 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done

File: server/libs/userlibs.js (hunk 5 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done

File: server/libs/userlibs.js (hunk 6 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done

File: server/libs/userlibs.js (hunk 7 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done

File: server/libs/userlibs.js (hunk 8 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done

File: server/routes/quiz.js (hunk 9 above)
Comment (HarisShahidHameed) [reply server/routes/quiz.js]: Done

File: server/routes/user.js (hunk 10 above)
Comment (HarisShahidHameed) [reply server/routes/user.js]: Done

File: server/routes/user.js (hunk 11 above)
Comment (HarisShahidHameed) [reply server/routes/user.js]: Done

File: server/routes/user.js (hunk 12 above)
Comment (HarisShahidHameed) [reply server/routes/user.js]: Done

**Review by abdulhaqshah — APPROVED** [review]

**Pull request merged**